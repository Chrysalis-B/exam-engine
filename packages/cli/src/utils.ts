import path from 'path'

export function resolveExam(filename: string) {
  const exam = filename.endsWith('.xml') ? filename : path.join(filename, 'exam.xml')
  return resolveFile(exam)
}

export function resolveFile(filename: string) {
  return path.resolve(process.cwd(), filename)
}

export function examName(exam: string) {
  return path.basename(path.dirname(exam))
}
