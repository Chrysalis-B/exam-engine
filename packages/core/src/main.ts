import '../src/css/main.less'

import Attachments from './components/Attachments'
import Exam from './components/Exam'
import parseExam from './parser/parseExam'

// This is the entry point for the exam UMD bundle (main-bundle.js)

export { Attachments, Exam, parseExam }
