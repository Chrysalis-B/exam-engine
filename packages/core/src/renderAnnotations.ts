import { Annotation, ImageAnnotation, TextAnnotation } from './types/Score'
import { createElement } from './dom-utils'
import classNames from 'classnames'

export function renderAnnotations(
  element: HTMLElement,
  pregradingAnnotations: Annotation[],
  censoringAnnotations: Annotation[]
): void {
  const annotations = [...pregradingAnnotations, ...censoringAnnotations]
  const annotationsWithMessages = annotations.filter((a) => a.message)

  for (const annotation of annotations) {
    const type = pregradingAnnotations.includes(annotation) ? 'pregrading' : 'censoring'
    const index = annotationsWithMessages.indexOf(annotation) + 1 || undefined

    switch (annotation.type) {
      case 'line':
      case 'rect':
        renderImageAnnotation(element, annotation, type, index)
        break
      default:
        renderTextAnnotation(element, annotation, type, index)
    }
  }
}

function renderImageAnnotation(
  element: HTMLElement,
  annotation: ImageAnnotation,
  type: 'pregrading' | 'censoring',
  index?: number
): void {
  const { attachmentIndex } = annotation
  const image = element.querySelectorAll('img')[attachmentIndex]

  const wrapper = getWrapper()
  const shape = mkShape()
  wrapper.appendChild(shape)

  function getWrapper() {
    const parent = image.parentElement!

    if (parent instanceof HTMLSpanElement) {
      return parent
    }

    const wrapper = createElement('span', { className: 'e-annotation-wrapper' })
    parent.insertBefore(wrapper, image)
    wrapper.appendChild(image)

    return wrapper
  }

  function mkShape() {
    const mark = createElement('mark', {
      className: classNames('e-annotation e-annotation--shape', {
        'e-annotation--pregrading': type === 'pregrading',
        'e-annotation--censoring': type === 'censoring',
      }),
      title: annotation.message,
    })

    const style = mark.style
    const pct = (n: number) => `${n * 100}%`

    if (annotation.type === 'rect') {
      style.left = pct(annotation.x)
      style.top = pct(annotation.y)
      style.right = pct(1 - (annotation.x + annotation.width))
      style.bottom = pct(1 - (annotation.y + annotation.height))
    } else {
      style.left = pct(annotation.x1)
      style.top = pct(annotation.y1)
      style.right = pct(1 - annotation.x2)
      style.bottom = pct(1 - annotation.y2)
    }

    if (index) {
      createSup(mark, index)
    }

    return mark
  }
}

function renderTextAnnotation(
  element: HTMLElement,
  annotation: TextAnnotation,
  type: 'pregrading' | 'censoring',
  index?: number
): void {
  const { startIndex, length: annotationLength } = annotation
  return go(element.childNodes[0], 0, annotationLength)

  function go(
    /* The current node in the answer HTML. */
    node: ChildNode,
    /** The current index corresponding to the node. */
    currentIndex: number,
    /* The remaining number of characters from the annotation to render. */
    remaining: number
  ): void {
    if (!isMark(node) && !isSup(node) && currentIndex >= startIndex) {
      // We found the correct spot to start an annotation. So let's create a mark element.
      const mark = mkMark()
      // Insert it before the node.
      node.parentElement?.insertBefore(mark, node)
      // Move the node and its next siblings inside the mark until the mark
      // contains the required amount of content or we run out of sibling nodes.
      // If the node is a Text node and it is too long for the annotation, we
      // split it in two and move the first part inside the mark, leaving the
      // rest in place.
      const stillRemaining = move(mark, node, remaining)

      // If we have run out of sibling nodes but still have characters to mark,
      // create another mark at the next text node. This means that marks have
      // been nested.
      if (stillRemaining > 0) {
        go(nextSibling(mark), currentIndex + annotationLength - stillRemaining, stillRemaining)
      }
      // We know that we're at the last mark of this annotation.
      // Render the superscript index after it, if necessary.
      else if (index) {
        createSup(mark, index)
      }
    } else {
      if (node instanceof Text && currentIndex + length(node) > startIndex) {
        // The annotation should start in the middle of this text node. We need to split it in two.
        node.splitText(startIndex - currentIndex)
      }

      go(next(node), currentIndex + length(node), remaining)
    }
  }

  function isMark(node: ChildNode): node is HTMLElement {
    return node.nodeName === 'MARK'
  }

  function isSup(node: ChildNode): node is HTMLElement {
    return node.nodeName === 'SUP'
  }

  function isImg(node: ChildNode): node is HTMLImageElement {
    return node.nodeName === 'IMG'
  }

  function next(node: ChildNode): ChildNode {
    return isMark(node) ? node.childNodes[0] : nextSibling(node)
  }

  function nextSibling(node: ChildNode): ChildNode {
    return node.nextSibling || nextSibling(node.parentElement!)
  }

  function length(node: ChildNode): number {
    return node instanceof Text ? node.length : isImg(node) ? 1 : 0
  }

  function mkMark(): HTMLElement {
    return createElement('mark', {
      className: classNames('e-annotation', {
        'e-annotation--pregrading': type === 'pregrading',
        'e-annotation--censoring': type === 'censoring',
      }),
      title: annotation.message,
    })
  }

  function move(mark: HTMLElement, node: ChildNode | null, remaining: number): number {
    if (!node || isMark(node) || remaining === 0) {
      return remaining
    }

    // Split text nodes if they are too long.
    if (node instanceof Text && length(node) > 0 && length(node) > remaining) {
      node.splitText(remaining)
    }

    const nextSibling = node.nextSibling
    mark.appendChild(node)
    return move(mark, nextSibling, remaining - length(node))
  }
}

function createSup(parent: Element, index: number) {
  const sup = createElement('sup', { className: 'e-annotation__index' }, `${index})`)
  parent.appendChild(sup)
  const width = sup.clientWidth
  // Write the new property in a rAF to avoid n reflows.
  requestAnimationFrame(() => {
    sup.style.right = `-${width + 4}px`
  })
}
