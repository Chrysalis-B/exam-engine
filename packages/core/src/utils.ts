/**
 * Maps an array, skipping the value if `fn` returns `undefined`.
 *
 * `mapMaybe(array, fn) â‰¡ array.map(fn).filter(v => v !== undefined)`
 */
export function mapMaybe<T, U>(array: T[], fn: (value: T, index: number) => U | undefined): U[] {
  const result: U[] = []

  for (let i = 0; i < array.length; i++) {
    const value = fn(array[i], i)
    if (value !== undefined) {
      result.push(value)
    }
  }

  return result
}

export function intersperse<T>(separator: T, array: T[]): T[] {
  const result = new Array<T>(array.length * 2 - 1)

  for (let i = 0; i < array.length; i++) {
    result[i * 2] = array[i]
    if (i !== array.length - 1) {
      result[i * 2 + 1] = separator
    }
  }

  return result
}
