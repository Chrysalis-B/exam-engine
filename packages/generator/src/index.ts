export * from './generateAnswers'
export * from './generateExam'
export * from './generateExamOptions'
