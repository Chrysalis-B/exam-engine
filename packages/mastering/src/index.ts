export { createMex, createMultiMex } from './createMex'
export * from './mastering'
export { ns } from './mastering/schema'
export { mkGetMediaMetadata } from './mkGetMediaMetadata'
