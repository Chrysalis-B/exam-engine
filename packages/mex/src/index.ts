export { createMex, createMultiMex } from './createMex'
export { getMediaMetadataFromLocalFile } from './getMediaMetadataFromLocalFile'
export * from './mastering'
