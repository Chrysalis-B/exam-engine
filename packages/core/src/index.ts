import '../src/css/main.less'
import Attachments from './components/attachments/Attachments'
import Exam from './components/exam/Exam'
import GradingInstructions from './components/grading-instructions/GradingInstructions'
import Grading from './components/grading/Grading'
import Results from './components/results/Results'
import parseExam from './parser/parseExam'

export { Attachments, Exam, Grading, GradingInstructions, Results, parseExam }
export { ExamComponentProps } from './createRenderChildNodes'
export * from './types/ExamAnswer'
export * from './types/ExamServerAPI'
export * from './types/GradingStructure'
export * from './types/Score'

export interface ExamBundle {
  Attachments: typeof Attachments
  Exam: typeof Exam
  parseExam: typeof parseExam
  Results: typeof Results
  Grading: typeof Grading
}
