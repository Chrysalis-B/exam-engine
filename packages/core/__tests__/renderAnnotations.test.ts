import { renderAnnotations } from '../src/renderAnnotations'
import { Annotation } from '../src'
import wrap from 'jest-snapshot-serializer-raw'

describe('renderAnnotations()', () => {
  it('one text annotation', () => {
    expect(
      render(
        'Lorem ipsum dolor sit amet',
        [
          {
            startIndex: 0,
            length: 5,
            message: 'foo',
          },
        ],
        []
      )
    ).toMatchSnapshot()
  })

  it('two text annotations, side by side', () => {
    expect(
      render(
        'Lorem ipsum dolor sit amet',
        [
          { startIndex: 0, length: 5, message: 'foo' },
          { startIndex: 5, length: 5, message: 'bar' },
        ],
        []
      )
    ).toMatchSnapshot()
  })

  it('two text annotations with space in between', () => {
    expect(
      render(
        'Lorem ipsum dolor sit amet',
        [
          { startIndex: 0, length: 5, message: 'foo' },
          { startIndex: 10, length: 5, message: 'bar' },
        ],
        []
      )
    ).toMatchSnapshot()
  })

  it('overlapping pregrading and censoring annotations', () => {
    expect(
      render(
        'Lorem ipsum dolor sit amet',
        [{ startIndex: 5, length: 10, message: 'foo' }],
        [{ startIndex: 0, length: 10, message: 'bar' }]
      )
    ).toMatchSnapshot()
  })

  it('text annotation through an image', () => {
    expect(
      render(
        'Lorem <img src="" /> ipsum dolor sit amet',
        [
          {
            startIndex: 0,
            length: 10,
            message: 'foo',
          },
        ],
        []
      )
    ).toMatchSnapshot()
  })

  it('rect annotation', () => {
    expect(
      render(
        `Lorem <img src="" /> ipsum dolor sit amet`,
        [
          {
            type: 'rect',
            attachmentIndex: 0,
            message: 'foo',
            x: 0.25,
            y: 0.25,
            height: 0.5,
            width: 0.5,
          },
        ],
        []
      )
    ).toMatchSnapshot()
  })

  it('line annotation', () => {
    expect(
      render(
        `Lorem <img src="" /> ipsum dolor sit amet`,
        [
          {
            type: 'line',
            attachmentIndex: 0,
            message: 'foo',
            x1: 0.25,
            y1: 0.25,
            x2: 0.75,
            y2: 0.25,
          },
        ],
        []
      )
    ).toMatchSnapshot()
  })
})

function createAnswer(html: string) {
  const div = document.createElement('div')
  div.innerHTML = html
  return div
}

function render(
  html = 'Lorem ipsum dolor sit amet',
  pregradingAnnotations: Annotation[],
  censoringAnnotations: Annotation[]
) {
  const answer = createAnswer(html)
  renderAnnotations(answer, pregradingAnnotations, censoringAnnotations)
  return wrap(answer.innerHTML)
}
