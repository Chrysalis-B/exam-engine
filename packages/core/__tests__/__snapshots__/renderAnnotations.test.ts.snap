// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderAnnotations() line annotation 1`] = `Lorem <span class="e-annotation-wrapper"><img src=""><mark class="e-annotation e-annotation--shape e-annotation--pregrading" title="foo" style="left: 25%; top: 25%; right: 25%; bottom: 75%;"><sup class="e-annotation__index">1)</sup></mark></span> ipsum dolor sit amet`;

exports[`renderAnnotations() one text annotation 1`] = `<mark class="e-annotation e-annotation--pregrading" title="foo">Lorem</mark><sup>1)</sup> ipsum dolor sit amet`;

exports[`renderAnnotations() overlapping pregrading and censoring annotations 1`] = `<mark class="e-annotation e-annotation--censoring" title="bar">Lorem</mark><mark class="e-annotation e-annotation--pregrading" title="foo"><mark class="e-annotation e-annotation--censoring" title="bar"> ipsu</mark><sup>2)</sup>m dol</mark><sup>1)</sup>or sit amet`;

exports[`renderAnnotations() rect annotation 1`] = `Lorem <span class="e-annotation-wrapper"><img src=""><mark class="e-annotation e-annotation--shape e-annotation--pregrading" title="foo" style="left: 25%; top: 25%; right: 25%; bottom: 25%;"><sup class="e-annotation__index">1)</sup></mark></span> ipsum dolor sit amet`;

exports[`renderAnnotations() text annotation through an image 1`] = `<mark class="e-annotation e-annotation--pregrading" title="foo">Lorem <img src=""> ip</mark><sup>1)</sup>sum dolor sit amet`;

exports[`renderAnnotations() two text annotations with space in between 1`] = `<mark class="e-annotation e-annotation--pregrading" title="foo">Lorem</mark><sup>1)</sup> ipsu<mark class="e-annotation e-annotation--pregrading" title="bar">m dol</mark><sup>2)</sup>or sit amet`;

exports[`renderAnnotations() two text annotations, side by side 1`] = `<mark class="e-annotation e-annotation--pregrading" title="foo">Lorem</mark><sup>1)</sup><mark class="e-annotation e-annotation--pregrading" title="bar"> ipsu</mark><sup>2)</sup>m dolor sit amet`;
